README.md
# rms-check

A linter and language server for Age of Empires 2 random map scripts.

[Usage](#usage) - [Status](#status) - [Install](#install) - [License: GPL-3.0](#license)

![VSCode integration](./screenshot.png)

## Usage

```
rms-check 0.1.0

USAGE:
    rms-check [FLAGS] [file] [SUBCOMMAND]

FLAGS:
        --aoc        Set the default compatibility to Age of Conquerors. Scripts can override this using `/*
                     Compatibility: */` comments.
    -h, --help       Prints help information
        --hd         Set the default compatibility to HD Edition. Scripts can override this using `/* Compatibility: */`
                     comments.
        --up14       Set the default compatibility to UserPatch 1.4. Scripts can override this using `/* Compatibility:
                     */` comments.
        --up15       Set the default compatibility to UserPatch 1.5. Scripts can override this using `/* Compatibility:
                     */` comments.
    -V, --version    Prints version information
        --wk         Set the default compatibility to WololoKingdoms. Scripts can override this using `/* Compatibility:
                     */` comments.

ARGS:
    <file>    The file to check, when not using any subcommand.

SUBCOMMANDS:
    check     Syntax check and lint a random map script.
    fix       Auto-fix problems with a random map script.
    help      Prints this message or the help of the given subcommand(s)
    pack      Pack a folder into an Zip-RMS map.
    server    Start the language server.
    unpack    Unpack a Zip-RMS map into a folder.
```

```bash
rms-check "/path/to/aoc/Random/Everything_Random_v4.3.rms"
```

## Status

There is a simple parser and some lints for highlighting common problems. There is also a barebones language server implementation that provides diagnostics and folding ranges.

In the future, I'd like to support more language server-y things, like automatic formatting, completions, and hover help.

## Install

Currently no binaries are provided. Installation must be done by compiling this repository from source.

First, get rustup: https://rustup.rs/

Then do something like:

```bash
# download rms-check:
git clone https://github.com/goto-bus-stop/rms-check.git
cd rms-check
# build it:
cargo build --release
# run it!
./target/release/rms-check FILENAME
```

## License

[GPL-3.0](./LICENSE.md)


vscode/CHANGELOG.md
# rms-check-vscode change log

All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](http://semver.org/).

## 0.0.4

### Language server
* Fix a crash when handling the `initialized` message.
* Implement `textDocument/definition`, providing "Go To Definition" support for consts.

### Command additions and fixes
* Add `set_gaia_unconvertible` and `set_gaia_civilization`.

### Known issues
* The optional "variance" argument to `circle_radius` is not yet supported.

## 0.0.3
Thanks to Chrazini and Zetnus for the feedback on the initial beta release!

### Command additions and fixes
* Add `enable_balanced_elevation`.
* Add layer attributes `base_layer` and `layer_to_place_on`.
* Add `avoid_cliff_zone`, `avoid_all_actor_areas`.
* Add `create_connect_to_nonplayer_land` and `default_terrain_replacement`.
* Remove extraneous argument type from `place_on_forest_zone`.

### Lint additions and fixes
* Update compatibility lint for DE update 36906.
* Allow using `nomad_resources` on HD Edition.

### Known issues
* The optional "variance" argument to `circle_radius` is not yet supported.

## 0.0.2
* Initial beta release.


src/main.rs
//! A syntax checker for Age of Empires 2 random map scripts.
//!
//! ```bash
//! rms-check "/path/to/aoc/Random/Everything_Random_v4.3.rms"
//! ```

mod check;
mod cli_reporter;
mod language_server;
mod zip_rms;

use crate::check::{cli_check, cli_fix, CheckArgs};
use crate::language_server::cli_server;
use crate::zip_rms::{cli_pack, cli_unpack};
use anyhow::Result;
use notify::{RecommendedWatcher, RecursiveMode, Watcher};
use rms_check::{Compatibility, FormatOptions};
use std::io::{self, Read};
use std::path::{Path, PathBuf};
use std::sync::mpsc;
use std::time::Duration;
use structopt::StructOpt;

// CLI flags for selecting a compatibility level.
#[derive(Debug, StructOpt)]
struct CliCompat {
    /// Set the default compatibility to Age of Conquerors. Scripts can override this using
    /// `/* Compatibility: */` comments.
    #[structopt(long = "aoc")]
    aoc: bool,
    /// Set the default compatibility to UserPatch 1.4. Scripts can override this using
    /// `/* Compatibility: */` comments.
    #[structopt(long = "up14")]
    userpatch14: bool,
    /// Set the default compatibility to UserPatch 1.5. Scripts can override this using
    /// `/* Compatibility: */` comments.
    #[structopt(long = "up15")]
    userpatch15: bool,
    /// Set the default compatibility to HD Edition. Scripts can override this using
    /// `/* Compatibility: */` comments.
    #[structopt(long = "hd")]
    hd_edition: bool,
    /// Set the default compatibility to WololoKingdoms. Scripts can override this using
    /// `/* Compatibility: */` comments.
    #[structopt(long = "wk")]
    wololo_kingdoms: bool,
    /// Set the default compatibility to Definitive Edition. Scripts can override this using
    /// `/* Compatibility: */` comments.
    #[structopt(long = "de")]
    definitive_edition: bool,
}

impl CliCompat {
    fn to_compatibility(&self) -> Compatibility {
        if self.definitive_edition {
            Compatibility::DefinitiveEdition
        } else if self.wololo_kingdoms {
            Compatibility::WololoKingdoms
        } else if self.hd_edition {
            Compatibility::HDEdition
        } else if self.userpatch14 {
            Compatibility::UserPatch14
        } else if self.userpatch15 {
            Compatibility::UserPatch15
        } else if self.aoc {
            Compatibility::Conquerors
        } else {
            Compatibility::All
        }
    }
}

#[derive(Debug, StructOpt)]
struct CliCheck {
    /// The file to check.
    file: PathBuf,
    #[structopt(flatten)]
    compat_flags: CliCompat,
}

#[derive(Debug, StructOpt)]
enum CliCommand {
    /// Pack a folder into an Zip-RMS map.
    #[structopt(name = "pack")]
    Pack {
        output: PathBuf,
        #[structopt(long, short = "d")]
        indir: PathBuf,
        #[structopt(long, short = "w")]
        watch: bool,
    },
    /// Unpack a Zip-RMS map into a folder.
    #[structopt(name = "unpack")]
    Unpack {
        #[structopt(long, short = "o")]
        outdir: PathBuf,
        input: PathBuf,
    },
    /// Auto-fix problems with a random map script.
    #[structopt(name = "fix")]
    Fix {
        /// Don't write the results.
        #[structopt(long = "dry-run")]
        dry_run: bool,
        /// The file to check.
        file: PathBuf,
        #[structopt(flatten)]
        compat_flags: CliCompat,
    },
    /// Format the given file.
    #[structopt(name = "format")]
    Format {
        /// The file to format. Use "-" to read from standard input.
        file: PathBuf,
        /// Set the size in spaces of a single tab indentation.
        #[structopt(long = "tab-size", default_value = "2")]
        tab_size: u32,
        /// Whether to use spaces instead of tabs for indentation.
        #[structopt(long = "no-use-spaces")]
        no_use_spaces: bool,
        /// Whether to align arguments in a list of commands.
        #[structopt(long = "no-align-arguments")]
        no_align_arguments: bool,
    },
    /// Syntax check and lint a random map script.
    #[structopt(name = "check")]
    Check(CliCheck),
    /// Start the language server.
    #[structopt(name = "server")]
    Server,
}

/// Syntax checking and linting tool suite for Age of Empires 2 random map scripts.
#[derive(Debug, StructOpt)]
#[structopt(name = "rms-check")]
pub struct Cli {
    #[structopt(subcommand)]
    command: Option<CliCommand>,
    // Compatibility flags for implicit `check`, when not using any subcommand.
    #[structopt(flatten)]
    compat_flags: CliCompat,
    /// The file to check, when not using any subcommand.
    file: Option<String>,
}

fn read_input(path: impl AsRef<Path>) -> io::Result<Vec<u8>> {
    let stdin = PathBuf::from("-");
    if path.as_ref() == stdin {
        let mut bytes = vec![];
        io::stdin().read_to_end(&mut bytes)?;
        Ok(bytes)
    } else {
        std::fs::read(path)
    }
}

/// Watch a directory for changes, and call the `callback` when something changes.
fn cli_watch(indir: impl AsRef<Path>, callback: impl Fn() -> Result<()>) -> Result<()> {
    callback()?;
    let (tx, rx) = mpsc::channel();
    let mut watcher: RecommendedWatcher = Watcher::new(tx, Duration::from_millis(500))?;
    watcher.watch(indir.as_ref(), RecursiveMode::Recursive)?;

    while let Ok(_event) = rx.recv() {
        match callback() {
            Ok(_) => (),
            Err(err) => {
                eprintln!("{}", err);
            }
        }
    }

    Ok(())
}

fn main() -> Result<()> {
    let args = Cli::from_args();

    match args.command {
        Some(CliCommand::Unpack { outdir, input }) => cli_unpack(input, outdir),
        Some(CliCommand::Pack {
            indir,
            output,
            watch,
        }) => {
            if watch {
                cli_watch(&indir, || {
                    cli_pack(&indir, &output)?;
                    println!("wrote {:?}", output);
                    Ok(())
                })
            } else {
                cli_pack(indir, output)
            }
        }
        Some(CliCommand::Fix {
            dry_run,
            file,
            compat_flags,
        }) => cli_fix(CheckArgs {
            compatibility: compat_flags.to_compatibility(),
            file,
            dry_run,
        }),
        Some(CliCommand::Format {
            file,
            tab_size,
            no_use_spaces,
            no_align_arguments,
        }) => {
            let options = FormatOptions::default()
                .tab_size(tab_size)
                .use_spaces(!no_use_spaces)
                .align_arguments(!no_align_arguments);

            let bytes = read_input(file)?;
            let string = std::str::from_utf8(&bytes)?;
            println!("{}", rms_check::format(string, options));
            Ok(())
        }
        Some(CliCommand::Server) => {
            cli_server();
            unreachable!();
        }
        Some(CliCommand::Check(args)) => cli_check(CheckArgs {
            compatibility: args.compat_flags.to_compatibility(),
            file: args.file,
            ..Default::default()
        }),
        None => {
            let args = CliCheck::from_args();
            cli_check(CheckArgs {
                compatibility: args.compat_flags.to_compatibility(),
                file: args.file,
                ..Default::default()
            })
        }
    }
}
